services/repository/branch.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repository
+
+import (
+	"errors"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	repo_module "code.gitea.io/gitea/modules/repository"
+	pull_service "code.gitea.io/gitea/services/pull"
+)
+
+// enmuerates all branch related errors
+var (
+	ErrBranchIsDefault   = errors.New("branch is default")
+	ErrBranchIsProtected = errors.New("branch is protected")
+)
+
+// DeleteBranch delete branch
+func DeleteBranch(doer *models.User, repo *models.Repository, gitRepo *git.Repository, branchName string) error {
+	if branchName == repo.DefaultBranch {
+		return ErrBranchIsDefault
+	}
+
+	isProtected, err := repo.IsProtectedBranch(branchName, doer)
+	if err != nil {
+		return err
+	}
+
+	if isProtected {
+		return ErrBranchIsProtected
+	}
+
+	commit, err := gitRepo.GetBranchCommit(branchName)
+	if err != nil {
+		return err
+	}
+
+	if err := gitRepo.DeleteBranch(branchName, git.DeleteBranchOptions{
+		Force: true,
+	}); err != nil {
+		return err
+	}
+
+	if err := pull_service.CloseBranchPulls(doer, repo.ID, branchName); err != nil {
```

Comment:
CloseBranchPulls is already called in PushUpdate here:

https://github.com/go-gitea/gitea/blob/ab77a24f18ae45248bd6c71778fe3bfacacf864f/services/repository/push.go#L216

Reply:
Yes, but it's safe to run it twice and there is slight performance reduce. The invoke in `push.go` will be invoked from both git command and `DeleteBranch` and others.